weedcoco/tests/testcases.py

```diff
@@ -0,0 +1,805 @@
+            "species": "daucus carota",
+            "eppo_taxon_code": "DAUCS",
+            "eppo_nontaxon_code": "3UMRC",
+            "role": "crop",
+            "id": 0,
+        },
+        {
+            "name": "weed: unspecified",
+            "species": "UNSPECIFIED",
+            "role": "weed",
+            "id": 1,
+        },
+    ],
+    "info": {
+        "description": "Cwfid annotations converted into WeedCOCO",
+        "metadata": {
+            "name": "Cwfid annotations converted into WeedCOCO",
+            "creator": [{"name": "Sebastian Haug"}],
+            "datePublished": "2015-XX-XX",
+            "license": "https://github.com/cwfid/dataset",
+        },
+    },
+    "license": [
+        {
+            "id": 0,
+            "url": "https://github.com/cwfid/dataset",
+        }
+    ],
+    "agcontexts": [
+        {
+            "id": 0,
+            "agcontext_name": "cwfid",
+            "crop_type": "daucus carota",
+            "bbch_growth_range": {"min": 10, "max": 20},
+            "soil_colour": "grey",
+            "surface_cover": "none",
+            "surface_coverage": "0-25",
+            "weather_description": "sunny",
+            "location_lat": 53,
+            "location_long": 11,
+            "camera_make": "JAI AD-130GE",
+            "camera_lens": "Fujinon TF15-DA-8",
+            "camera_lens_focallength": 15,
+            "camera_height": 450,
+            "camera_angle": 90,
+            "camera_fov": 22.6,
+            "photography_description": "Mounted on boom",
+            "ground_speed": 0,
+            "lighting": "natural",
+            "cropped_to_plant": False,
+        }
+    ],
+}
+
+test_bad_category_name_expected = [
+    {
+        "error_type": "jsonschema",
+        "n_errors_found": "1",
+        "error_details": [
+            {
+                "path": ["categories", 0, "name"],
+                "value": "foobar",
+                "message": "'foobar' does not match '^((crop|weed): ([a-z][a-z ]+|UNSPECIFIED)|crop|weed|none)$'",
+                "schema": {
+                    "type": "string",
+                    "pattern": "^((crop|weed): ([a-z][a-z ]+|UNSPECIFIED)|crop|weed|none)$",
+                    "description": 'A category for an annotation, consisting of an agricultural role and\noptionally a biological name, separated by ": ". Currently allowed\nagricultural roles are "weed" and "crop".  The "none" category may be\nused to indicate that no weed or crop is identified in the image.\n\nAnnotations should be labelled with the most specific category possible,\nusually "weed: <species name>", but super categories may be more general.\nBiological names should be lowercase.\nFor an unspecified species of weed, "weed: UNSPECIFIED" may be used.\n\nFor example, to label an oat crop in an image, the category name would be\n"crop: avena sativa". Its super categories may include "crop: grasses"\nand "crop". Oats as a volunteer crop would be "weed: avena sativa".',
```

Comment:
I think it's a bad idea to reproduce the schema here, since small changes to the schema will need to be replicated in the test suite. They're not essential to testing our code either.

I don't have a simple solution except to break up the assertion... I struggled with this internally, so I've had a hack at how I might write that:

```py
def check_error_details(actual, expected):
    def _everything_but_details(obj):
        return ({k: v for k, v in obj.items() if k != "error_details"}

    def _stable_details(obj):
        return ({k: v for k, v in obj["error_details"].items() if k not in ["schema", "message"]}

    assert actual.keys() == expected.keys()
    assert _everything_but_details(actual) == _everything_but_details(expected)
    assert actual["error_details"].keys() == expected["error_details"].keys()
    assert _stable_details(actual) == _stable_details(expected)
```

Reply:
I think it's easier just to remove the schema from actual, as the structure of it can be predicted